plone/recipe/pound/docs/configuring.txt
Configuring
***********

Supported options
=================

The recipe supports the following options:

daemon - Daemon 0|1
    Have Pound run in the foreground (if 0) or as a daemon (if 1). By default Pound runs as a daemon (detaches itself from the controlling terminal and puts itself in the background). By specifying this option you can force Pound to work like a regular process. Useful for debugging or if you want to use something like daemontools.

log_facility - LogFacility value
    Specify the log facility to use. value (default: daemon) must be one of the symbolic facility names defined in syslog.h. This facility shall be used for logging. Using a - for the facility name causes Pound to log to stdout/stderr.

log_level - LogLevel value
    Specify the logging level: 0 for no logging, 1 (default) for regular logging, 2 for extended logging (show chosen backend server as well), 3 for Apache-like format (Combined Log Format with Virtual Host), 4 (same as 3 but without the virtual host information) and 5 (same as 4 but with information about the Service and BackEnd used). This value can be overridden for specific listeners.

log_file - a log file location
    This option is only meaningful when the log_facility is "-". In this case, the output from pound will be redirected to this file.

dynscale - DynScale 0|1
    Enable or disable the dynamic rescaling code (default: 0). If enabled Pound will periodically try to modify the back-end priorities in order to equalise the response times from the various back-ends. This value can be overridden for specific services.

alive - Alive value
    Specify how often Pound will check for resurected back-end hosts (default: 30 seconds). In general, it is a good idea to set this as low as possible - it will find resurected hosts faster. However, if you set it too low it will consume resources - so beware.

client - Client value
    Specify for how long Pound will wait for a client request (default: 10 seconds). After this long has passed without the client sending any data Pound will close the connection. Set it higher if your clients time-out on a slow network or over-loaded server, lower if you start getting DOS attacks or run into problems with IE clients. This value can be overridden for specific listeners.

timeout - TimeOut value
    How long should Pound wait for a response from the back-end (in seconds). Default: 15 seconds.

grace - Grace value
    How long should Pound continue to answer existing connections after a receiving and INT or HUP signal (default: 30 seconds). The configured listeners are closed immediately. You can bypass this behaviour by stopping Pound with a TERM or QUIT signal, in which case the program exits without any delay.

balancers - list of listener http with backend
    Set the list of backends, the list by default is
    **name adress:port adress-backend-1:port-backend-1(,priority,timeOut)  adress-backend-2:port-backend-2(,priority,timeOut) ... adress-backend-n:port-backend-n(,priority,timeOut)**

executable - path to pound, required if pound executable is not provided by the build recipe

socket - Set the control socket path.

sticky - if on , add a sticky session configuration (default on, possible value on/off)

sessiontype - Type of the sticky session (default COOKIE)

sessioncookie - Name of the cookie of the stickysession (default __ac)

sessiontimeout - Time out of the sticky session (default 300)

poundctl-binary - Set the init script name
poundrun-binary - Set the daemon binary name
poundcontrol-binary - Set the pound controller name  (it wraps the pound's poundctl binary)

Example usage
=============

We'll start by creating a buildout that uses the recipe::

    >>> write('buildout.cfg',
    ... """
    ... [buildout]
    ... parts = configpound
    ...
    ... [configpound]
    ... recipe = plone.recipe.pound:config
    ... executable = /usr/sbin/pound
    ... balancers =
    ...    one  127.0.0.1:80 127.0.0.1:8080 127.0.0.1:8081,1
    ...    two  0.0.0.0:81 127.0.0.1:8082,1,120 127.0.0.1:8083 169.1.1.2:80
    ...
    ... """ )

Running the buildout gives us::

    >>> print system(buildout)
    Installing configpound.
    <BLANKLINE>



CHANGES.txt
trunk (2011-05-21)
==================

  - added log_file option to allow output redirection to this file,
    when using with log_facility set to -
    [Tiberiu Ichim]
  - only set ID in session block of pound.cfg if value of sessiontype is 
    COOKIE; otherwise Pound will not start (e.g., ID is set but type
    is IP) [bryanw]
  - xxx [Ingeniweb]

0.5.5 (2009-06-02)
==================

  - cmmi change extra_options configuration [youenn]

0.5.4 (2008-12-12)
==================

  - Add new options to configure sticky session [youenn]


0.5.3 (2008-10-23)
==================

  - fix regression in fetching current group [youenn]
  - poundctl restart need to use the right config file as well [jensens]
  - In term of portability, bash scripts now use env to get their shell.
  - The poundctl script overrides the control binary present in $prefix/sbin.
    So are introduced options to fix that problem: (kiorky)

        - A `poundcontrol` script is added to wrap the real poundctl binary.
        - Customize the names of generated scripts.
          To use them, simply add  `poundctl-binary`, `poundrun-binary`,
          `poundcontrol-binary` to your options.

0.5.1 (2008-07-08)
==================

  - fix bug in fecthing current group [youenn]

0.5.0 (2008-07-07)
==================

  - rename iw.recipe.pound to plone.recipe.pound [youenn]
  - fix doctests [youenn]
  - add socket option [mathieu pasquet]
  - config recipe now additionally creates "poundrun" script
    which merely runs configured pound executable with correct configuration
    [rockyburt]
  - fetching current running user is now more robust [rockyburt]


0.4.1 (2008-07-02)
==================

  - fix Cheetah require [youenn]

0.4.0 (2008-06-26)
==================

  - add a ctl script (as plone.recipe.squid) [youenn]

0.3.0
=====

  - devide recipe in two part : build and config part [youenn]
  - add options to build (extra-options) [youenn]
  - add options to the config file (time-out, priority ...) [youenn]

0.2.0
=====

 - added a script to launch pound with the right conf file [tarek]

0.1.0
=====

 - Initial release [tarek]


plone/recipe/pound/config.py
# -*- coding: utf-8 -*-
# Copyright (C)2008 'ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Config Recipe pound"""

import os
import stat
import zc
from types import BooleanType
from Cheetah.Template import Template

from plone.recipe.pound import utils

RECIPE_BUILD_NAME = 'plone.recipe.pound:build'


def get_options_from_build(buildout, options):
    part = options.get('poundbuildpart', None)
    if part:
        return buildout[part]

    for part in buildout.keys():
        if buildout[part].has_key('recipe') and \
            buildout[part]['recipe'] == RECIPE_BUILD_NAME:
            return buildout[part]
    return {}


class ConfigureRecipe(object):
    """This recipe is used by zc.buildout"""

    def __init__(self, buildout, name, options):
        self.buildout = buildout
        location = options.get(
            'location', buildout['buildout']['parts-directory'])
        options['location'] = os.path.join(location, name)
        options['prefix'] = options['location']

        self.options = options
        self.buildoptions = get_options_from_build(buildout, options)
        self.options['socket'] = self.buildoptions.get(
            'socket',
            os.path.join(
                self.buildout['buildout']['directory'],
                'var',
                'pound.sock'
            )
        )

    def install(self):
        """ install config fo pound """

        # configures pound
        # loading templates
        curdir = os.path.dirname(__file__)
        tpl = Template(open(os.path.join(curdir, 'pound.cfg.tpl')).read())
        try:
            tpl.daemon = int(self.options.get('daemon',1))
        except ValueError:
            raise zc.buildout.UserError("Deamon is invalid")
        tpl.log_facility = self.options.get('log_facility','daemon')
        try:
            tpl.log_level = int(self.options.get('log_level',1))
        except ValueError:
            raise zc.buildout.UserError("LogLevel is invalid")
        try:
            tpl.alive = int(self.options.get('alive',30))
        except ValueError:
            raise zc.buildout.UserError("Alive is invalid")
        try:
            tpl.dynscale = int(self.options.get('dynscale',0))
        except ValueError:
            raise zc.buildout.UserError("Dynscale is invalid")

        try:
            tpl.client = int(self.options.get('client',10))
        except ValueError:
            raise zc.buildout.UserError("Client is invalid")
        try:
            tpl.timeout = int(self.options.get('timeout',15))
        except ValueError:
            raise zc.buildout.UserError("Timeout is invalid")

        try:
            tpl.grace = int(self.options.get('grace',30))
        except ValueError:
            raise zc.buildout.UserError("Grace is invalid")
        try:
            tpl.owner = self.buildoptions.get('owner', utils.get_sysuser())
        except ValueError:
            raise zc.buildout.UserError("Owner is invalid")
        try:
            tpl.group = self.buildoptions.get('group', utils.get_sysgroup())
        except ValueError:
            raise zc.buildout.UserError("Group is invalid")
        ## session management
        try:
            tpl.sticky = self.options.get('sticky', 'on')
        except ValueError:
            raise zc.buildout.UserError("Sticky is invalid")
        try:
            tpl.sessiontype = self.options.get('sessiontype', 'COOKIE')
        except ValueError:
            raise zc.buildout.UserError("Sessiontype is invalid")
        try:
            tpl.sessioncookie = self.options.get('sessioncookie', '__ac')
        except ValueError:
            raise zc.buildout.UserError("Sessioncookie is invalid")
        try:
            tpl.sessiontimeout = int(self.options.get('sessiontimeout', 300))
        except ValueError:
            raise zc.buildout.UserError("SessionTimeout is invalid")

        tpl.socket = self.options.get('socket', '')

        # creating balancers
        balancer_cfg = []

        for balancer in self.options.get('balancers', '').split('\n'):

            balancer = balancer.strip()
            if balancer == '':
                continue
            balancer = balancer.split()
            name = balancer[0]
            try:
                (adress, port) = balancer[1].split(':')
            except ValueError:
                raise zc.buildout.UserError("balancer syntax is not correct %s" \
                                            % balancer)

            backends = balancer[2:]
            backends_cfg = []

            for backend in backends:
                backend = backend.split(':')
                host = backend[0]
                port_backend = backend[1]
                priority = 0
                timeout = 0
                if ',' in port_backend:
                    l =  port_backend.split(',')
                    if len(l) == 3:
                        (port_backend, priority, timeout) = l

                    elif len(l) == 2:
                        (port_backend, priority) = l
                    else:
                        raise zc.buildout.UserError("Backend configuration is invalid")
                backends_cfg.append({'host': host,
                                    'port': port_backend,
                                    'priority' : priority,
                                    'timeout': timeout})

            balancer_dict = {'name': name,
                             'port': port,
                             'adress': adress,
                             'backends': backends_cfg}

            balancer_cfg.append(balancer_dict)
        tpl.balancers = balancer_cfg


        # writing the file
        if not  os.path.exists(self.options['location']):
            os.mkdir(self.options['location'])
        etc_dir = os.path.join(self.options['location'], 'etc')
        if not os.path.exists(etc_dir):
            os.mkdir(etc_dir)
        var_dir = os.path.join(self.options['location'], 'var')
        if not os.path.exists(var_dir):
            os.mkdir(var_dir)

        filename = os.path.join(etc_dir, 'pound.cfg')
        f = open(filename, 'w')

        try:
            print >>f, tpl
        finally:
            f.close()

        if self.buildoptions:
            self.options['executable'] = os.path.join(self.buildoptions['location'],
                                                      'sbin','pound')
        else:
            if not self.options.get('executable'):
                raise zc.buildout.UserError("path to executable is required")


        pid = os.path.join(var_dir, 'pound.pid')
        scripts = []
        for x in ('poundctl', 'poundrun', 'poundcontrol'):
            ctl = Template(open(os.path.join(curdir, x+'.tpl')).read())
            ctl.poundbin = self.options['executable']
            ctl.socket = self.options.get('socket', '')
            ctl.poundcontrol = os.path.join(
                os.path.dirname(self.options['executable']),
                'poundctl'
            )
            ctl.logfilename = self.options.get('log_file', os.path.join(var_dir, 'pound.log'))
            ctl.poundcfg = filename
            ctl.poundpid = pid
            bin_dir = self.buildout['buildout']['bin-directory']
            script_name = os.path.join(bin_dir, self.options.get('%s-binary' % x, x))
            f = open(script_name, 'w')
            try:
                print >>f, ctl
            finally:
                f.close()
            mode = os.stat(script_name).st_mode
            os.chmod(script_name, mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
            scripts.append(script_name)

        return [etc_dir] + scripts
